Add support for saving multiple wifi networks

SZenglein opened the pull request:
With this branch, it is now possible to save up to 10 different SSID-Password combinations.

I basically reworked the complete Wlan implementation. IMHO it looks cleaner and easier to understand now using discrete states. Tell me what you think.
(There are a lot of Serial.print statements. I plan to remove those eventually.)

It works pretty well as is, but could have unintended side effects. Looking for testers. 
**Running this will delete your WIFI settings and migrate them to the new NVS keys.**

What works:
* The new code scans available networks
* Tries connecting to each known network, starting at the strongest one
* Up to 10 SSIDs are saved in NVS
* SSIDs can be added and removed
* **For now, the old ui is used to add networks, so you can already test adding multiple networks**
* go to http://espuino.local/savedSSIDs to view your ssids

What does not work:
* mainly the frontend is incomplete
* No way to trigger the delete network function
* HTML does not yet list ssids

biologist79 commented:
Didn't review everything so far... but as a general advice: please use ESPuino's logging. Always!

SZenglein commented:
Of course, the prints are for development and most will be removed completely.

laszloh reviewed (changes requested):
So all in all I like the PR. 

There is one point, which is global I did not know where to put is `knownNetworks` and the handling of the settings. It feels clumpsy how it's used currently (like an idear started but never finshed). We basically fetch all the saved networks from NVS into an array to use it once in a single function.

The idear of putting the data into the RAM is good (since NVS is slow), but currently it just feels like we are wasting memory.

So I'd propose to either only use NVS (and live with the slowdown) or to make the SSID in the RAM the master (and mirror it out into the NVS). Then we should also make use of the RAM entries more often.

<br>

If we take route 2, we could also think in a next step about putting more than the SSID into the RAM. I thought about an array of structures like this:
```cpp
struct SavedNetwork {
    char SSID[32];   // or String
    char pass[64];   // or String
    IPAddress ip;    // if this is 0.0.0.0 then we are dynamic (or introduce a bool into the struct)
    IPAddress subment;
    IPAddress gateway;
};
```
This structure is 120 bytes, so we'd need 1,2kB to put them into the RAM (56 bytes + dynamic memory for the actual SSID&PWD strings, if we use String). 

These could be serialized for NVS storage relatively easily and hold all the info we need (so we'd have 1 NVS entry / Network and not 2 independent ones).

SZenglein commented on src/Wlan.cpp:
```diff
@@ -10,173 +10,481 @@
 
-	if (accessPointStarted && dnsServer) {
-		dnsServer->processNextRequest();
+bool Wlan_DeleteNetwork(String ssid) {
```
You're right. I'm not sure what happened here, but the count should obviously only be decreased on success.

Right now it does not matter that only the NVS is updated because the variable is seldomly used, but I agree it's bad design.

SZenglein replied:
fixed in 1d3c9f6

I'm not a fan of the index based erase. I thought of it initally, but IMO a stable key is better.

tueddy commented on src/Web.cpp:
```diff
@@ -1166,6 +1220,92 @@ void explorerHandleAudioRequest(AsyncWebServerRequest *request) {
+		StaticJsonDocument<8192> jsonBuffer;
+	#endif
+	String serializedJsonString;
+	JsonObject obj = jsonBuffer.createNestedObject();
```
The issue is a bit stuck, although it makes so much sense! 
Can we do something to move it forward?

SZenglein replied:
Yeah, sorry. @laszloh provided very valuable feedback and I did plan on reworking at least the NVS storage of credentials.
But time was a bit short and I didn't get to make the changes yet, but I hope to do so soon.

SZenglein replied:
In response to the original comment: I'm pretty sure this is not returning a nested object. "obj" is an object with one key.
You could argue returning a single string is better though.

SZenglein commented on src/Web.cpp:
```diff
@@ -1166,6 +1220,92 @@ void explorerHandleAudioRequest(AsyncWebServerRequest *request) {
+
+void handleGetHostname(AsyncWebServerRequest *request) {
+	String hostname = Wlan_GetHostname();
+	request->send(200, (char *) F("text/plain; charset=utf-8"), hostname);
```
Right, that shouldn't work. Thanks JS.

I'm not convinced sending json is the best way for just a single string. But maybe it's better to merge this with "getActiveSSID" into a json object

SZenglein commented:
> 
> ```c++
> struct SavedNetwork {
>     char SSID[32];   // or String
>     char pass[64];   // or String
>     IPAddress ip;    // if this is 0.0.0.0 then we are dynamic (or introduce a bool into the struct)
>     IPAddress subment;
>     IPAddress gateway;
> };
> ```

> These could be serialized for NVS storage relatively easily and hold all the info we need (so we'd have 1 NVS entry / Network and not 2 independent ones).


In 1d3c9f6d935bc9992903463bb5351bf93fe22b5f I changed how it's handled to your proposal.

I added a very similar struct. I then went a step further and store the whole array of networks in one NVS entry. This entry is loaded only once at startup and written if something changes.

SZenglein commented on src/Wlan.cpp:
```diff
@@ -10,173 +10,481 @@
+uint8_t connectionAttemptCounter = 0;
+unsigned long connectStartTimestamp = 0;
+uint32_t connectionFailedTimestamp = 0;
+String knownNetworks[10];
```
Changed in 1d3c9f6d935bc9992903463bb5351bf93fe22b5f 

Let me know if you think it's better

SZenglein commented on src/Wlan.cpp:
```diff
@@ -10,173 +10,481 @@
+char* getSSIDKey(uint8_t index) { 
+	snprintf(nvs_key, 9, "SSID_%d", index);
+	return nvs_key;
+}
```
this is not relevant anymore

SZenglein commented on src/Wlan.cpp:
```diff
@@ -10,173 +10,481 @@
+bool Wlan_DeleteNetwork(String ssid) {
+	uint8_t len = numSavedNetworks;
+	numSavedNetworks--;
+	gPrefsSettings.putUChar("NUM_NETWORKS", numSavedNetworks);
```
resolving because outdated

SZenglein commented on src/Wlan.cpp:
```diff
@@ -10,173 +10,481 @@
+	}
+}
+
+bool Wlan_SetHostname(String newHostname) {
```
Explain this function in arduino-espressif then, hah!
```
bool WiFiGenericClass::setHostname(const char * hostname)
{
    set_esp_netif_hostname(hostname);
    return true;
}
```

tueddy commented:
@SZenglein  Thank's for your work! Let me know if you're ready, then i test again.
Please try to resolve conflicts with current DEV-branch so merging is possible..

SZenglein commented on src/Wlan.cpp:
```diff
@@ -10,173 +10,481 @@
-			#endif
+	if (initialStart) {
+		initialStart = false;
+		accessPointStart((char *) FPSTR(accessPointNetworkSSID), (char *) FPSTR(accessPointNetworkPassword), apIP, apNetmask);
```
I didn't change this from the dev branch, but I'll remove it now....

SZenglein commented:
> @SZenglein Thank's for your work! Let me know if you're ready, then i test again. Please try to resolve conflicts with current DEV-branch so merging is possible..

Feel free to test again. I personally don't have a setup for testing the static ip configuration, so that part is completely untested by me.

tueddy commented:
I've tested again with my Wifi + phone hotspot and it work's fine!
I will test in office with a third router tomorrow & merge then..

SZenglein commented:
> I've tested again with my Wifi + phone hotspot and it work's fine! I will test in office with a third router tomorrow & merge then..
> 
> The compiler gives these warnings, can this be avoided?
> 
> src/Wlan.cpp: In function 'bool Wlan_DeleteNetwork(String)': src/Wlan.cpp:466:91: warning: 'void* memmove(void*, const void*, size_t)' writing to an object of type 'struct WiFiSettings' with no trivial copy-assignment; use copy-assignment or copy-initialization instead [-Wclass-memaccess] memmove(knownNetworks+i, knownNetworks+i+1, sizeof(WiFiSettings)*(numKnownNetworks-i-1)); ^ In file included from src/Wlan.cpp:13: src/Wlan.h:4:8: note: 'struct WiFiSettings' declared here struct WiFiSettings { ^~~~~~~~~~~~

I think the warning might be because the compiler thinks IPAddress is not plain old bytes. Not sure if it's actually illegal to copy that class.
